File: sdk/typescript/src/policy.ts
import type { AgentPayClient } from "./client";
import type {
  PolicyConfig,
  PolicyResponse,
  IssueSessionKeyRequest,
  SessionKeyResponse,
} from "./types";

/**
 * Set the spending policy for a smart account (R8.1, R11.1).
 *
 * PUT /v1/policy/:smartAccount
 *
 * @param client - The AgentPay API client
 * @param smartAccount - The smart account address
 * @param config - Per-transaction and daily spending caps
 * @returns The updated policy
 */
export async function setPolicy(
  client: AgentPayClient,
  smartAccount: string,
  config: PolicyConfig,
): Promise<PolicyResponse> {
  return client.put<PolicyResponse>(`/v1/policy/${smartAccount}`, config);
}

/**
 * Issue a new session key (R13.1, R11.1).
 *
 * POST /v1/policy/session-keys
 *
 * @param client - The AgentPay API client
 * @param request - Session key details including keyId, publicKey, window, bounds
 * @returns The created session key
 */
export async function issueSessionKey(
  client: AgentPayClient,
  request: IssueSessionKeyRequest,
): Promise<SessionKeyResponse> {
  return client.post<SessionKeyResponse>("/v1/policy/session-keys", request);
}

/**
 * Revoke a session key (R13.3, R11.1).
 *
 * DELETE /v1/policy/session-keys/:keyId
 *
 * @param client - The AgentPay API client
 * @param keyId - UUIDv7 identifier of the session key to revoke
 */
export async function revokeSessionKey(
  client: AgentPayClient,
  keyId: string,
): Promise<void> {
  await client.delete(`/v1/policy/session-keys/${keyId}`);
}

File: sdk/python/src/agentpay_sdk/policy.py
"""Policy management (mirrors sdk/typescript/src/policy.ts)."""

from dataclasses import asdict

from .client import AgentPayClient
from .types import PolicyConfig, PolicyResponse, IssueSessionKeyRequest, SessionKeyResponse


def set_policy(
    client: AgentPayClient,
    smart_account: str,
    config: PolicyConfig,
) -> PolicyResponse:
    """Set the spending policy for a smart account.

    PUT /v1/policy/:smartAccount
    """
    data = client.put(f"/v1/policy/{smart_account}", asdict(config))
    return PolicyResponse(
        smart_account=data["smartAccount"],
        per_tx_cap_usdc_micro=data["perTxCapUsdcMicro"],
        daily_cap_usdc_micro=data["dailyCapUsdcMicro"],
        rolling_24h_spend_usdc_micro=data["rolling24hSpendUsdcMicro"],
        remaining_daily_usdc_micro=data["remainingDailyUsdcMicro"],
        updated_at=data["updatedAt"],
    )


def issue_session_key(
    client: AgentPayClient,
    request: IssueSessionKeyRequest,
) -> SessionKeyResponse:
    """Issue a new session key.

    POST /v1/policy/session-keys
    """
    body = {
        "keyId": request.key_id,
        "smartAccount": request.smart_account,
        "publicKey": request.public_key,
        "notBefore": request.not_before,
        "notAfter": request.not_after,
        "bounds": asdict(request.bounds),
    }
    data = client.post("/v1/policy/session-keys", body)
    return SessionKeyResponse(
        key_id=data["keyId"],
        smart_account=data["smartAccount"],
        public_key=data["publicKey"],
        not_before=data["notBefore"],
        not_after=data["notAfter"],
        bounds=data["bounds"],
        status=data["status"],
        issued_at=data["issuedAt"],
        revoked_at=data.get("revokedAt"),
    )


def revoke_session_key(client: AgentPayClient, key_id: str) -> None:
    """Revoke a session key.

    DELETE /v1/policy/session-keys/:keyId
    """
    client.delete(f"/v1/policy/session-keys/{key_id}")
